test_conformance/spirv_new/test_op_spec_constant.cpp

```diff
@@ -0,0 +1,182 @@
+                       T spec_constant_value,
+                       bool (*notEqual)(const T &, const T &) = isNotEqual<T>)
+{
+    if (std::string(name).find("double") != std::string::npos)
+    {
+        if (!is_extension_available(deviceID, "cl_khr_fp64"))
+        {
+            log_info("Extension cl_khr_fp64 not supported; skipping double "
+                     "tests.\n");
+            return TEST_SKIPPED_ITSELF;
+        }
+    }
+    if (std::string(name).find("half") != std::string::npos)
+    {
+        if (!is_extension_available(deviceID, "cl_khr_fp16"))
+        {
+            log_info(
+                "Extension cl_khr_fp16 not supported; skipping half tests.\n");
+            return TEST_SKIPPED_ITSELF;
+        }
+    }
+
+    clProgramWrapper prog;
+    cl_int err = get_program_with_il(prog, deviceID, context, name);
+    SPIRV_CHECK_ERROR(err, "Failed to build program");
+
+    clKernelWrapper kernel = clCreateKernel(prog, "spec_const_kernel", &err);
+    SPIRV_CHECK_ERROR(err, "Failed to create kernel");
+    int num = (int)references.size();
+    size_t bytes = num * sizeof(T);
+    clMemWrapper numbers_buffer =
+        clCreateBuffer(context, CL_MEM_READ_WRITE | CL_MEM_USE_HOST_PTR, bytes,
+                       inputs.data(), &err);
+    SPIRV_CHECK_ERROR(err, "Failed to create numbers_buffer");
+
+    clMemWrapper spec_buffer =
+        clCreateBuffer(context, CL_MEM_USE_HOST_PTR,
+                       sizeof(spec_constant_value), &spec_constant_value, &err);
+    SPIRV_CHECK_ERROR(err, "Failed to create spec_buffer");
+
+    err = clSetKernelArg(kernel, 0, sizeof(clMemWrapper), &numbers_buffer);
+    SPIRV_CHECK_ERROR(err, "Failed to set kernel argument inputs_buffer");
+    err = clSetKernelArg(kernel, 1, sizeof(clMemWrapper), &spec_buffer);
+    SPIRV_CHECK_ERROR(err, "Failed to set kernel argument spec_constant_value");
+
+    size_t global = num;
+    err = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, &global, NULL, 0, NULL,
+                                 NULL);
+    SPIRV_CHECK_ERROR(err, "Failed to enqueue kernel");
+    clFinish(queue);
+
+    std::vector<T> device_results(num, 0);
+    err = clEnqueueReadBuffer(queue, numbers_buffer, CL_TRUE, 0, bytes,
+                              device_results.data(), 0, NULL, NULL);
+    SPIRV_CHECK_ERROR(err, "Failed to copy from numbers_buffer");
+
+    for (int i = 0; i < num; i++)
+    {
+        if (device_results[i] == references[i])
+        {
+            log_error("Values match but should not at location %d expected %d "
+                      "obtained %d\n",
+                      i, references[i], device_results[i]);
+            return TEST_FAIL;
+        }
+    }
+
+    err = clSetProgramSpecializationConstant(prog, 101, sizeof(T),
+                                             &spec_constant_value);
+    SPIRV_CHECK_ERROR(err, "Failed to run clSetProgramSpecializationConstant");
+
+    err = clBuildProgram(prog, 1, &deviceID, NULL, NULL, NULL);
```

Comment:
This can fail with CL_INVALID_OPERATION, because there are already kernel objects attached to the program. Please use a new program, _without_ a call to `clBuildProgram()` between the call to `clCreateProgramWithIL()` and the call to `clSetProgramSpecializationConstant()`.

Reply:
Could you clarify? My understanding is that changing value by clSetProgramSpecializationConstant requires program rebuilding using clBuildProgram. You can read about this [here](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_setting_spir_v_specialization_constants). Noticed that clSetProgramSpecializationConstant is run second time.

Reply:
In my opinion there is no need to create program object once again. However I create once again kernel object and clSetKernelArg calls are executed

Reply:
> Could you clarify? My understanding is that changing value by clSetProgramSpecializationConstant requires program rebuilding using clBuildProgram. You can read about this [here](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_setting_spir_v_specialization_constants). Noticed that clSetProgramSpecializationConstant is run second time.

`clSetProgramSpecializationConstant()` must be called _before_ `clBuildProgram()`. `clBuildProgram()` is called by `get_program_with_il()`.

> In my opinion there is no need to create program object once again.

The [specification states](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clBuildProgram) that `clBuildProgram()` returns:
> * `CL_​INVALID_​OPERATION` if there are kernel objects attached to _program_.

The test must not call `clBuildProgram()` for a program after calling `clCreateKernel()` for that same program.

> However I create once again kernel object and clSetKernelArg calls are executed

It is not clear to me what you are saying, here. Are you simply describing what the code in the PR currently does? If so, I've read the code and understand what is written: there's no real need to turn it into English, for me, at least.

Reply:
Looks that second usage the objects I've made is forbidden.  Then I will create new objects. Thanks